### ISSUE TITLE
systemd-networkd: per-interface-type route metrics

### ISSUE THREAD

**Issue by flokli** [systemd/systemd#17698]
Apparently, systemd-networkd defaults to a route metric of 1024 for all interfaces, while NetworkManager defaults to 600 for wireless and 300 for wired (IIRC) - not sure about wwan/LTE interfaces.

I use `systemd-networkd` to manage wired interfaces, and NetworkManager to manage wireless interfaces (by setting `unmanaged-devices` in their config)

When using networkd without any custom `RouteMetric=` set, this will result in my system preferring wifi over wired, which feels quite unintuitive.

IMHO, networkd should try to use the same metrics here by default.

**Comment by yuwata** [comment]
Hmm, why we need to adjust the value, although the value is already configurable?

**Comment by flokli** [comment]
This is already configurable, yes.
My proposal here would be instead of defaulting to 1024 in all cases, defaulting to per-interface-type default (like NetworkManager does when you don't configure metrics explicitly).

 - People with a single interface don't care about metrics anyways, as that's a single default route
 - People with multiple interfaces don't need to juggle with metrics by themselves anymore, and networkd will have a preference to "more wired" interfaces out of the box (like NetworkManager has)
 - People with NetworkManager and networkd co-existing will not suddenly have default traffic unexpectedly being preferred to go over the wireless interface

**Comment by yuwata** [comment]
These make some sense. However, changing the default value easily breaks backward compatibility... I think it is hard to do that.

**Comment by flokli** [comment]
I'd argue most people who didn't explicitly configure `RouteMetric=` simply didn't care, as it didn't matter for their usecase - So now having another metric default here are very unlikely to break things.

If my usecase would cover multiple different route metrics, I wouldn't just configure it for one connection, and leave networkd to create the remaining ones with metric 1024,but always set them explicitly.

So networkd changing defaults here shouldn't break much. Of course, it should be noted in the changelog, but IMHO it'd be an improvement and make things much nicer out of the box.

**Comment by RPigott** [comment]
I think this is would be a reasonable change.

I doubt there are many users if any that specifically rely on the default route metric, especially since the defaults happen to be the same for all connections. I'm not actually sure how linux selects between routes with equivalent metric.

**Comment by yuwata** [comment]
But, for example, suppose that a user set routes on a wifi adapter with metric 800, and on a wired interface with the default metric (1024). If we change the default value to one used NetworkManager used, 600 for wifi and 300 for wired, then the preference of the user's machine become inverted, 800 for wifi and 300 (new default) for wired... Such an unwanted backward incompatibility sounds unacceptable for that user.

Also, not only NetworkManager and systemd-networkd, there are many daemons or tools to manage network interfaces exist. I guess they have their own default values. What happens if we use dhcpcd for a wifi interface and networkd for a wired interface (though I do not know what is the default value in dhcpcd)? Even if we change the default value as suggested by @flokli, still we may need to adjust the metrics for such the case. I cannot find any strong reason that networkd need to use the values used by NetworkManager.

**Comment by flokli** [comment]
Yes, it's possible to construct cases where there would be a change in behaviour, if we expect people relied on the 1024 default metric for one connection, and explicitly configured a metric for another connection.

However, as explained above, the chance of this being the case is probably very very small, and having per-interface-type metrics would really be a huge increase in convenience for all the other cases.

Using the same values as NetworkManager would make these metrics quasi-defaults, and would also keep the same "out of the box route metric behaviour" if for example Fedora would switch from NetworkManager to networkd by default, so I think defaulting on these values makes sense.

**Comment by poettering** [comment]
I sympathize with the issue, and I also see the compat breakage this might introduce. I am tempted to say though we should do this nevertheless, and bite the bullet. I have the suspicion breakage would be minor. Of course, we should announce this loudly in NEWS to give people a heads-up.

**Comment by kerbless** [comment]
I can say that from the experience of a new arch user wanting to try iwd + systemd-networkd to manage the network, this part of the setup was somewhat surprising and would have been hard to spot if it wasn't for the  archwiki [citing this issue](https://wiki.archlinux.org/title/Systemd-networkd#Wired_and_wireless_adapters_on_the_same_machine).
So for my context the change makes total sense.
